File: Server/utils/Parity.java
package Server.utils;

/**
 * Parity class provides methods to calculate and verify even parity bits.
 * It is used to ensure data integrity during serialization and deserialization.
 */
public class Parity {
    /**
     * Calculate the even parity bit for a byte array.
     * 
     * @param data The byte array to calculate parity for.
     * @return 1 if the total number of 1 bits is odd, 0 if it is even.
     */
    public static byte calculateEvenParityBit(byte[] data) {
        byte parityBit = 0;

        // XOR all bytes together to get odd/even count of 1 bits
        for (byte b : data) {
            // Process each bit in the byte
            byte temp = b;
            for (int i = 0; i < 8; i++) {
                // XOR the lowest bit with our running parity
                parityBit ^= (temp & 1);
                temp >>= 1;
            }
        }

        // For even parity, if parityBit is 1, we need to add a 1 to make total 1s even
        return parityBit;
    }

    /**
     * Verify the even parity for a byte array and a provided parity bit.
     * 
     * @param data      The byte array to verify.
     * @param parityBit The parity bit to check against.
     * @return True if the parity is valid, false otherwise.
     */
    public static boolean verifyEvenParity(byte[] data, byte parityBit) {
        // Calculate what the parity should be
        byte calculatedParity = calculateEvenParityBit(data);

        // For even parity, the calculated parity should match the provided parity bit
        return calculatedParity == parityBit;
    }
}

File: Client/src/Parity.cpp
#include "Parity.hpp"

/**
 * @brief Calculates the even parity bit for a given set of data.
 * 
 * The even parity bit is calculated such that the totla number of 1s in the data (including the parity bit) is even.
 * 
 * @param data A vector of bytes for which the parity bit is to be calculated.
 * 
 * @return The calculated even parity bit (0 or 1).
 */
uint8_t Parity::calculateEvenParityBit(const std::vector<uint8_t> &data)
{
    uint8_t parityBit = 0;
    for (const auto &byte : data)
    {
        uint8_t temp = byte;
        while (temp)
        {
            parityBit ^= (temp & 1); // XOR the least significant bit
            temp >>= 1; // Shift right to process the next bit
        }
    }
    return parityBit;
}

/**
 * @brief Verifies the even parity of a given set of data and a parity bit.
 * 
 * This function caluclates the evan parity bit for the given data and compares it with the provided parity bit to determine if the parity is correct.
 * 
 * @param data A vector of bytes for which the parity is to be verified.
 * @param parityBit The provided parity bit to verify against.
 * 
 * @return True if the calculated parity matches the provided parity bit, false otherwise.
 */
bool Parity::verifyEvenParity(const std::vector<uint8_t> &data, uint8_t parityBit)
{
    uint8_t calculatedParity = calculateEvenParityBit(data);
    return calculatedParity == parityBit;
}
